File: CPlusPlus/1776_Car_Fleet_II.cpp
class Solution {
public:
    vector<double> getCollisionTimes(vector<vector<int>>& cars) {
        int N = cars.size();
        vector<int> dists;
        vector<int> vels;

        for (auto& car : cars) {
            dists.push_back(car[0]);
            vels.push_back(car[1]);
        }
        vector<double> result(N, -1), stack;

        auto collisionTime = [&](int i, int j) {
            return (double)(dists[i] - dists[j]) / (vels[j] - vels[i]);
        };

        for (int i = N - 1; i >= 0; i--) {
            while (
                (!stack.empty() && vels[i] <= vels[stack.back()]) ||
                (stack.size() >= 2 && collisionTime(i, stack.back()) > result[stack.back()])) {
                stack.pop_back();
            }
            if (!stack.empty()) result[i] = collisionTime(i, stack.back());
            stack.push_back(i);
        }
        return result;
    }
};

File: Python/1776_Car_Fleet_II.py
class Solution:
    def getCollisionTimes(self, cars: List[List[int]]) -> List[float]:
        N = len(cars)
        dists, vels = zip(*cars)
        result = [-1] * N
        stack = []
        
        def collision_time(i, j):
            diff_dist = dists[i] - dists[j]
            diff_vel = vels[j] - vels[i]
            return diff_dist / diff_vel if diff_vel else float('inf')
        
        for i in range(N -1, -1, -1):
            while (
                stack and vels[i] <= vels[stack[-1]] or 
                len(stack) >= 2 and collision_time(i, stack[-1]) > collision_time(stack[-1], stack[-2])
            ):
                stack.pop()
        
            if stack:
                result[i] = collision_time(i, stack[-1])
            stack.append(i)
        return result
